### DISCUSSION TITLE
route search params type not working when it's a union and have parent search params

### DISCUSSION THREAD

**Discussion by craxrev** [TanStack/router#2977] [Q&A] [answered]
Hey, so let's say I have the following `validateSearch`:
```typescript
  validateSearch: z.union([
    z.object({
      test11: z.string(),
      test12: z.string(),
    }),
    z.object({
      test13: z.string(),
      test14: z.string(),
    }),
  ]),
```
Now, when using `Route.useSearch()`, there's two different behaviours of the returned type depending on the parent search params:
1. when the parent have search params: no types, just parent search params
2. when parent doesn't have search params: expected types, following the schema above, it'd be:
```typescript
const test: {
    test11: string;
    test12: string;
} | {
    test13: string;
    test14: string;
}
```
On runtime, it behaves as expected, meaning it will respect the schema provided and will throw errors accordingly, it's just the typescript types that are behaving incorrectly, so I'm not sure this is a bug or that we're not supposed to use unions..

**Comment by craxrev** [comment]
I found that having the following `validateSearch` works:
```typescript
  validateSearch: z.union([
    z.object({
      test11: z.string(),
      test12: z.string(),
      test13: z.never().optional(),
      test14: z.never().optional(),
    }),
    z.object({
      test11: z.never().optional(),
      test12: z.never().optional(),
      test13: z.string(),
      test14: z.string(),
    }),
  ]),
```
Now, search params with the above schema will get merged as expected with parent search params, so if for example the parent have a schema like so: `z.object({ redirect: z.string().optional() })`, the resulting type would be then:
```typescript
const test: {
    redirect?: string | undefined;
    test11: string;
    test12: string;
    test13?: undefined;
    test14?: undefined;
} | {
    redirect?: string | undefined;
    test13: string;
    test14: string;
    test11?: undefined;
    test12?: undefined;
}
```
before it was only:
```typescript
const test: {
    redirect?: string | undefined;
}
```

**Reply by schiller-manuel** [reply]
can you please provide a complete minimal example by forking one of the existing examples on stackblitz?

**Reply by craxrev** [reply]
yeah, here's the stackblitz [link](https://stackblitz.com/edit/tanstack-router-papvyeqa?file=src%2Froutes%2F_layout%2F_layout-2.tsx), `layout-a.tsx` will have the `validateSearch` with the issue, `layout-b.tsx` will have the working one.

**Comment by schiller-manuel** [comment, marked as answer]
https://github.com/TanStack/router/releases/tag/v1.87.12 should fix this

**Reply by craxrev** [reply]
Thanks, works now

**Reply by craxrev** [reply]
Now I come to think about it, I find the type with "optional undefined" properties easier to use, instead of always narrowing the type first in order to use any "non-common" properties.. or is there any other way to use it?

Can't we just make the "optional undefined" the default behaviour? or better, is there a zod utility that does that, specially when I don't want to retype all my "non-common" properties on each object type of the union (think of the case of a union with more than 2 object types)
